packages/shared/src/api/lanyardApi.ts

```diff
@@ -82,32 +92,91 @@ export async function discoverContacts(phoneNums: string[]) {
     try {
       await pokeNoun({ app: 'lanyard', mark: 'lanyard-query-1', noun });
     } catch (e) {
-      console.error('contact discovery poke error', e);
-    }
-    const queryResponse = await queryResponseSub;
-    console.log(`bulk result`, queryResponse);
-
-    if (queryResponse) {
-      // return matches
-      const nextSalt = queryResponse.result?.['next-salt'];
-      const matches = queryResponse.result?.result
-        ? Object.entries(queryResponse.result.result).filter(([_key, value]) =>
-            Boolean(value)
-          )
-        : [];
-
-      return matches;
+      logger.trackError('contact discovery poke error', e);
+      logger.trackEvent(AnalyticsEvent.DebugContactMatching, {
+        error: e,
+        errorMessage: e.message,
+      });
     }
+    try {
+      const queryResponse = await queryResponseSub;
+
+      if (queryResponse && queryResponse.query.nonce === encodedNonce) {
+        if (queryResponse.query.result === 'rate limited') {
+          logger.log('rate limited');
+          logger.trackEvent(AnalyticsEvent.DebugContactMatching, {
+            error: 'rate limited',
+            errorMessage: 'rate limited',
+          });
+          return [];
+        }
+
+        const nextSalt = queryResponse.query.result?.['next-salt'];
+        await db.lastLanyardSalt.setValue(nextSalt);
+        const matches = queryResponse.query.result?.results
+          ? (Object.entries(queryResponse.query.result.results).filter(
+              ([_key, value]) => Boolean(value)
+            ) as [string, string][])
+          : [];
 
-    return [];
+        return matches;
+      }
+
+      return [];
+    } catch (e) {
+      logger.error('error in discoverContacts', e);
+      logger.trackEvent(AnalyticsEvent.DebugContactMatching, {
+        parser: 'discoverContacts',
+        error: e,
+        errorMessage: e.message,
+      });
+      throw e;
+    }
   } catch (e) {
-    console.error('error discovering contacts', e);
+    logger.error('error discovering contacts', e);
     throw e;
   }
 }
 
-function diffContactBook(phoneNums: string[]) {
-  // TODO: diff last request
+async function diffContactBook(phoneNums: string[]) {
+  const last = await db.lastPhoneContactSetRequest.getValue();
+  logger.log('last phone contact set', last);
+
+  if (last) {
+    const lastSet = JSON.parse(last);
+    logger.log('lastSet', lastSet);
+
+    // find new phone numbers (additions)
+    const diff = phoneNums.filter((num) => !lastSet.includes(num));
+    logger.log('diff', diff);
+
+    // find removed phone numbers (deletions)
+    const delSet = lastSet.filter((num: string) => !phoneNums.includes(num));
+    logger.log('delSet', delSet);
+
+    // always store the CURRENT complete set of phone numbers
```

Comment:
how does immediately updating the stored set work if the actual `%verifier` request fails? 

If we try again, will there be an issue with previous salt not matching passed `last/add/del`? Would we incur a big hit to our rate limit balance?

Reply:
Good point, we could only write it out after we successfully hear back from the verifier.